src/envoy/http/service_control/client_cache.cc

```diff
@@ -381,37 +388,50 @@ void ClientCache::handleCheckResponse(const Status& http_status,
     }
   } else {
     if (translate_non_5xx) {
+      // Most likely an auth error in ESPv2 or API producer deployment.
+      filter_stats_.filter_.denied_producer_error_.inc();
+
       // This is not caused by a client request error, so translate
       // non-5xx error codes to 500 Internal Server Error. Error message
       // contains details on the original error (including the original
       // HTTP status code).
       Status scrubbed_status(Code::INTERNAL, final_status.error_message());
       on_done(scrubbed_status, response_info);
     } else {
+      // Stats already incremented for this case.
       on_done(final_status, response_info);
     }
   }
   delete response;
 }
 
-void ClientCache::callQuota(
-    const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
-    std::function<void(const ::google::protobuf::util::Status& status)>
-        on_done) {
+void ClientCache::callQuota(const AllocateQuotaRequest& request,
+                            QuotaDoneFunc on_done) {
   auto* response = new AllocateQuotaResponse;
-  client_->Quota(
-      request, response, [this, response, on_done](const Status& status) {
-        if (status.ok()) {
-          on_done(::espv2::api_proxy::service_control::RequestBuilder::
-                      ConvertAllocateQuotaResponse(*response,
-                                                   config_.service_name()));
-        } else {
-          on_done(Status(static_cast<google::protobuf::util::error::Code>(
-                             status.error_code()),
-                         status.error_message()));
-        }
-        delete response;
-      });
+  client_->Quota(request, response,
+                 [this, response, on_done](const Status& status) {
+                   handleQuotaResponse(status, response, on_done);
+                 });
+}
+
+void ClientCache::handleQuotaResponse(const Status& http_status,
+                                      AllocateQuotaResponse* response,
+                                      QuotaDoneFunc on_done) {
+  if (http_status.ok()) {
+    QuotaResponseInfo response_info;
+    Status quota_status = ::espv2::api_proxy::service_control::RequestBuilder::
+        ConvertAllocateQuotaResponse(*response, config_.service_name(),
+                                     &response_info);
+
+    handleScResponseErrorStats(response_info.error_type);
+    on_done(quota_status);
+  } else {
+    // Most likely an auth error in ESPv2 or API producer deployment.
```

Comment:
@qiwzhang: This branch should never occur, status should always be OK for Quota calls. I'll remove it in a different PR, I would like to add some integration tests to verify.

Reply:
this only happen when cache is disabled.  But we don't support disabling quota cache.  With cache,  the rate limiting is not very precious,   e.g. if limit is 300 p/m,   we may go 500 p/m.  So maybe some users may perfer long latency, but precious limit.

Reply:
Ack. I feel no customers have such strict requirements, but we can follow-up later.